// @jd/upload-oss-tools/dist_cli/model/downloadTarget.js
"use strict";
/*
 * @Author: zhuzhida
 * @Date: 2020-07-20 13:22:18
 * @Last Modified by: zhuzhida
 * @Last Modified time: 2021-06-08 16:19:29
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadTarget = void 0;
/**
 *
 * 下载目录表
 * @class downloadTarget
 */
var downloadTarget = /** @class */ (function () {
    function downloadTarget() {
        /**
         *
         * 随机时间戳唯一id
         * @type {string}
         * @memberof downloadTarget
         */
        this._id = "";
        /**
         *
         * 空间名称
         * @type {string}
         * @memberof downloadTarget
         */
        this.bucket = "";
        /**
         *
         * 本地存储路径
         * @type {string}
         * @memberof downloadTarget
         */
        this.downloadLocalFullPath = "";
        /**
         *
         * 是否映射远程绝对路径
         * @type {boolean}
         * @memberof downloadTarget
         */
        this.remoteFolder = true;
        /**
         *
         * 开始下载 回调函数
         * @type {function}
         * @memberof uploadTarget
         */
        this.downloadStart = function () { };
        /**
         *
         * 下载中 回调函数
         * @type {function}
         * @memberof downloadTarget
         */
        this.downloadProgress = function () { };
        /**
         *
         * 下载完成 回调函数
         * @type {function}
         * @memberof downloadTarget
         */
        this.downloadEnd = function () { };
    }
    return downloadTarget;
}());
exports.downloadTarget = downloadTarget;
