// @jd/upload-oss-tools/dist_cli/model/uploadTarget.js
"use strict";
/*
 * @Author: zhuzhida
 * @Date: 2020-07-20 13:22:18
 * @Last Modified by: zhuzhida
 * @Last Modified time: 2020-09-23 14:19:39
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadTarget = void 0;
/**
 *
 * 上传目录表
 * @class uploadTarget
 */
var uploadTarget = /** @class */ (function () {
    function uploadTarget() {
        /**
         *
         * 随机时间戳唯一id
         * @type {string}
         * @memberof uploadTarget
         */
        this._id = "";
        /**
         *
         * 空间名称
         * @type {string}
         * @memberof uploadTarget
         */
        this.bucket = "";
        /**
         *
         * 本地完整绝对路径
         * @type {string}
         * @memberof uploadTarget
         */
        this.localFullPath = "";
        /**
         *
         * 排除的文件
         * @type {string}
         * @memberof uploadTarget
         */
        this.ignoreRegexp = "";
        /**
         *
         * 是否覆盖上传
         * @type {boolean}
         * @memberof uploadTarget
         */
        this.cover = true;
        /**
         *
         * 是否下载 刷新cdn文件日志
         * @type {boolean}
         * @memberof uploadTarget
         */
        this.printCdnFile = false;
        /**
         *
         * 开始上传 回调函数
         * @type {function}
         * @memberof uploadTarget
         */
        this.uploadStart = function () { };
        /**
         *
         * 上传中 回调函数
         * @type {function}
         * @memberof uploadTarget
         */
        this.uploadProgress = function () { };
        /**
         *
         * 上传完成 回调函数
         * @type {function}
         * @memberof uploadTarget
         */
        this.uploadEnd = function () { };
    }
    return uploadTarget;
}());
exports.uploadTarget = uploadTarget;
